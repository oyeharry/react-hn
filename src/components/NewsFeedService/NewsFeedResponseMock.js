/* eslint-disable  */
export default {
  hits: [
    {
      created_at: '2011-10-05T23:42:23.000Z',
      title: 'Steve Jobs has passed away.',
      url: 'http://www.apple.com/stevejobs/',
      author: 'patricktomas',
      points: 4271,
      story_text: '',
      comment_text: null,
      num_comments: 376,
      story_id: null,
      story_title: null,
      story_url: null,
      parent_id: null,
      created_at_i: 1317858143,
      relevancy_score: 3496,
      _tags: ['story', 'author_patricktomas', 'story_3078128'],
      objectID: '3078128',
      _highlightResult: {
        title: {
          value: 'Steve Jobs has passed away.',
          matchLevel: 'none',
          matchedWords: [],
        },
        url: {
          value: 'http://www.apple.com/stevejobs/',
          matchLevel: 'none',
          matchedWords: [],
        },
        author: { value: 'patricktomas', matchLevel: 'none', matchedWords: [] },
        story_text: { value: '', matchLevel: 'none', matchedWords: [] },
      },
    },
    {
      created_at: '2017-02-19T21:16:33.000Z',
      title: 'Reflecting on one very, very strange year at Uber',
      url:
        'https://www.susanjfowler.com/blog/2017/2/19/reflecting-on-one-very-strange-year-at-uber',
      author: 'grey-area',
      points: 4107,
      story_text: null,
      comment_text: null,
      num_comments: 1014,
      story_id: null,
      story_title: null,
      story_url: null,
      parent_id: null,
      created_at_i: 1487538993,
      relevancy_score: 7260,
      _tags: ['story', 'author_grey-area', 'story_13682022'],
      objectID: '13682022',
      _highlightResult: {
        title: {
          value: 'Reflecting on one very, very strange year at Uber',
          matchLevel: 'none',
          matchedWords: [],
        },
        url: {
          value:
            'https://www.susanjfowler.com/blog/2017/2/19/reflecting-on-one-very-strange-year-at-uber',
          matchLevel: 'none',
          matchedWords: [],
        },
        author: { value: 'grey-area', matchLevel: 'none', matchedWords: [] },
      },
    },
  ],
  nbHits: 20659005,
  page: 1,
  nbPages: 500,
  hitsPerPage: 2,
  exhaustiveNbHits: false,
  query: '',
  params:
    'advancedSyntax=true\u0026analytics=true\u0026analyticsTags=backend\u0026hitsPerPage=2\u0026page=1\u0026tags=%28story%29',
  processingTimeMS: 3,
};
