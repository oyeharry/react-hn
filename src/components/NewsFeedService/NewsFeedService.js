import fetch from 'cross-fetch';
import formatDistanceStrict from 'date-fns/formatDistanceStrict';

const algoliaApiEndpoint = 'https://hn.algolia.com/api/v1';
const newsFeedSearchEndpoint = `${algoliaApiEndpoint}/search?tags=story`;
const newsFeedSearchByDateEndpoint = `${algoliaApiEndpoint}/search_by_date?tags=story`;
const storyUrls = new Map([
  ['topstories', newsFeedSearchEndpoint],
  ['newstories', newsFeedSearchByDateEndpoint],
]);

function getProcessedNewsFeedData(newsFeedData) {
  const { hits: newsFeedHits = [], page, nbPages, hitsPerPage } = newsFeedData;
  return {
    page,
    nbPages,
    hitsPerPage,
    hits: newsFeedHits.map(newsFeedHit => {
      const {
        author: username,
        num_comments: totalComments,
        title,
        url = '',
        created_at_i: createdAtI,
        points: totalUpVotes,
        objectID,
      } = newsFeedHit;

      const [linkDomain] =
        (url && url.match(/(?!w{1,}\.)(\w+\.?)([a-zA-Z]+)(\.\w+)/)) || [];

      const postedWhen = `${formatDistanceStrict(
        createdAtI * 1000,
        Date.now()
      )} ago`;

      return {
        totalComments,
        totalUpVotes,
        title,
        linkDomain,
        username,
        postedWhen,
        url,
        id: objectID,
      };
    }),
  };
}

function getTimestampOfDays(day) {
  return new Date().setDate(new Date().getDate() - day) / 1000;
}

export async function queryNewsFeed(params) {
  const {
    pageNum = 0,
    hitsPerPage = 30,
    dateRange = 'last24h',
    storyType = 'topstories',
  } = params;
  let filters;

  if (dateRange === 'last24h') {
    filters = `&numericFilters=created_at_i>${getTimestampOfDays(1)}`;
  }

  const apiUrl = storyUrls.get(storyType);
  const url = `${apiUrl}&page=${pageNum}&hitsPerPage=${hitsPerPage}${filters}`;

  try {
    const response = await fetch(url);
    if (response.ok) {
      const newsFeedData = await response.json();
      return getProcessedNewsFeedData(newsFeedData);
    } else {
      return { error: true };
    }
  } catch (e) {
    return { error: true };
  }
}

export default { queryNewsFeed };
