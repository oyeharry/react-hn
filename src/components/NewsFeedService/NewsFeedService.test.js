import fetch from 'cross-fetch';

jest.mock('cross-fetch', () => jest.fn());

import newsFeedMockResponse from './NewsFeedResponseMock';
import NewsFeedService from './NewsFeedService';

const mockedResponse = Promise.resolve({
  json: () => {
    return {
      ...newsFeedMockResponse,
    };
  },
});

describe('NewsFeedService', () => {
  it('Should get valid response of news feed', async () => {
    fetch.mockImplementation(() => mockedResponse);
    const response = await NewsFeedService.queryNewsFeed(1);
    expect(response).toMatchSnapshot();
  });
});
