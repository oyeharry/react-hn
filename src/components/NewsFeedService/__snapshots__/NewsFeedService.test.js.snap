// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewsFeedService Should get valid response of news feed 1`] = `
Object {
  "hits": Array [
    Object {
      "id": "3078128",
      "linkDomain": "apple.com",
      "postedWhen": "9 years ago",
      "title": "Steve Jobs has passed away.",
      "totalComments": 376,
      "totalUpVotes": 4271,
      "url": "http://www.apple.com/stevejobs/",
      "username": "patricktomas",
    },
    Object {
      "id": "13682022",
      "linkDomain": "susanjfowler.com",
      "postedWhen": "3 years ago",
      "title": "Reflecting on one very, very strange year at Uber",
      "totalComments": 1014,
      "totalUpVotes": 4107,
      "url": "https://www.susanjfowler.com/blog/2017/2/19/reflecting-on-one-very-strange-year-at-uber",
      "username": "grey-area",
    },
  ],
  "hitsPerPage": 2,
  "nbPages": 500,
  "page": 1,
}
`;
